### PR TITLE
fix a bug: elpy-find-file does not prune enough

### PR DESCRIPTION
**Pull request by sam-s**
(elpy-project-ignored-directories): extend the standard `vc-directory-exclusion-list`

(elpy-ffip-prune-patterns): extend `ffip-prune-patterns` using
`elpy-project-ignored-directories` and `completion-ignored-extensions`

(elpy-find-file): bind `ffip-prune-patterns` to `elpy-ffip-prune-patterns`

### CODE REVIEW

**Comment by jorgenschaefer** [comment]
Hello and thanks for the contribution! Is there a reason for having two variables there? It seems kind of redundant, especially as changing one won't affect the relevant one until Elpy is reloaded. Either have only one variable, or calculate the correct value dynamically when ffip is called.

Also, could you look at that test error? I think it's related to your code?

**Comment by sam-s** [comment]
elpy-project-ignored-directories is used in several places.
its use in connection with ffip is wrong.

it is not clear to me that re-setting elpy-ffip-prune-patterns when elpy-project-ignored-directories is modified is TRT.

**Comment by sam-s** [comment]
I think I fixed the test failure.

**Comment by sam-s** [comment]
I am confused by the single remaining failure - it appears that `cask` fails on parsing a TAR file.
I don't see how it can be my fault, and I don't know how to force re-checking.

**Comment by jorgenschaefer** [comment]
Looks like that was a temporary problem with MELPA. Restarting the build fixed it, it seems. Thank you!

**Comment by sam-s** [comment]
any chance you might consider merging this?
thanks!

**Comment by jorgenschaefer** [comment]
I am afraid I have been rather busy the last weeks. I will look at this when I have more time. Sorry for the inconvenience.

**Comment by jorgenschaefer** [comment]
Hello, and sorry for the delay. I'm confused by this PR – there are three commits that seem unrelated to each other, and they also seem to have been against the master branch so they get polluted by my own changes there. Could you please fix the PR so it's only the patch you want me to look at?

**Comment by sam-s** [comment]
Hi, the 3 patches are small and I suggest that you look at all of them.
They should be fairly uncontroversial.

File: elpy.el [hunk 1]
```diff
@@ -92,9 +92,9 @@ can be inidividually enabled or disabled."
                      elpy-module-sane-defaults))
   :group 'elpy)
 
-(defcustom elpy-project-ignored-directories ' (".bzr" "CVS" ".git" ".hg" ".svn"
-                                               ".tox"  "build" "dist"
-                                               ".cask")
+(defcustom elpy-project-ignored-directories
+  (append '(".tox" "build" "dist" ".cask")
+          vc-directory-exclusion-list)
```
Comment (jorgenschaefer) [inline elpy.el]: Nice! :-)

File: elpy.el [hunk 2]
```diff
@@ -1335,6 +1335,23 @@ for."
...
+    (dolist (dir elpy-project-ignored-directories)
+      (let ((pat (concat "*/" dir "/*")))
+        (unless (member pat ffip-prune-patterns)
+          (push pat extras))))
+    (dolist (ext completion-ignored-extensions)
+      (unless (string-match "/$" ext)
+        (let ((pat (concat "*" ext)))
+          (unless (member pat ffip-prune-patterns)
+            (push pat extras)))))
+    (append extras ffip-prune-patterns))
```
Comment (jorgenschaefer) [inline elpy.el]: Initializing one defcustom on the basis of another leads to user confusion. It's a good idea to have a function that merges the two defcustoms instead. Also, more readable.

**Comment by sam-s** [comment]
nested ifs have their place.
I used them for a reason - to avoid re-computing the same function multiple times.

**Comment by sam-s** [comment]
I pushed the changes you requested.

**Comment by jorgenschaefer** [comment]
The `(require 'cl-extra)` line seems to break on the CI.

**Comment by sam-s** [comment]
looks like it works now!

**Comment by sam-s** [comment]
what is holding the patches at this time?

**Comment by jorgenschaefer** [comment]
> what is holding the patches at this time?

Me not having the time to review stuff all the time.

Could you please squash the commits? I'll merge this PR then.

**Comment by sam-s** [comment]
I don't think commits that have already been pushed can be squished.

**Comment by jorgenschaefer** [comment]
> I don't think commits that have already been pushed can be squished.

`git push --force` works :-)

**Comment by sam-s** [comment]
how do you want me to squish?
1. http://stackoverflow.com/a/9254257/850781
1. http://stackoverflow.com/a/23486788/850781

**Comment by jorgenschaefer** [comment]
I do not really care _how_ you do it. The point of doing this is to clean up the history so it does show the semantic changes, not the intermediate steps you did to arrive there. Especially a [readable git commit message](http://chris.beams.io/posts/git-commit/) is important there.

It might be that your commits consist of two or three semantically different units. Then those can stay separate. Otherwise, just squash them all together.

Personally, I usually use `git rebase -i` to clean up branches. For larger stuff, I sometimes use `git reset ; git add .` – matter of personal preference.

**Comment by sam-s** [comment]
You know, I am sorry, on a second thought I am not sure I am prepared to jump through more hoops.
I offered a few bug fixes 3 months ago - and complied with your code reviews - and even have your LGTM above.
Now you are asking for money (presumably as a joke, but the suggestion that your time is more valuable than mine is ridiculous - you don't know me, I don't know you, I see no reason to think either way) and git squishes. Why do I need git squishes after your LGTM?

**Comment by jorgenschaefer** [comment]
Asking you to squash (squash, not squish, it's a technical term) commits is not making you "jump through hoops", it's standard practice in many projects. Sorry if this is new to you. You'll benefit a lot personally by learning about it.

And I did not ask for money from you. I informed you that if you do not like my response times, I am willing to make a special exception just for you, personally, as a special snowflake, to get my full and undivided attention the moment you request it, for a standard fee. Assuming you do not want to pay that (which is fine), I am afraid you have to wait _like everyone else_ for me to find the time to respond to you.

**Comment by sam-s** [comment]
Fixes https://github.com/jorgenschaefer/elpy/issues/1102

**Comment by jorgenschaefer** [comment]
Considering [this comment](https://github.com/jorgenschaefer/elpy/issues/1102#issuecomment-289529243), I assume my requests (squash related commits; create multiple PRs for unrelated changes) won't be followed, so as I won't merge this PR as is, I'll close it.

Thank you for your contribution anyhow. Please feel free to create new PRs if and when you are willing to not make my work unnecessarily annoying.

**Pull request closed without merging**